File: Sean的leetcode/python/3Sum Smaller.py
###
# Two Pointer
# Time Complexity: O(nlogn) + O(n^2)
# Space Complexity: O(1)
###
class Solution(object):
    def threeSumSmaller(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        if len(nums) < 3:
            return 0
        nums.sort()
        cnt = 0
        for i in xrange(len(nums) - 2):
            l = i + 1
            r = len(nums) - 1
            while l < r:
                s = nums[i] + nums[l] + nums[r]
                if s < target:
                    cnt += (r-l)
                    l += 1
                else:
                    r -= 1
        return cnt
        
        

File: MyLeetCode/Leetcode/3Sum Smaller.cpp
//https://leetcode.com/discuss/52442/o-n-2-c-solution
class Solution {
public:
    int threeSumSmaller(vector<int>& nums, int target) {
        int n = nums.size();
        if (n < 3) return 0;
        sort(nums.begin(), nums.end());
        int count = 0;
        for (int i = 0; i < n - 2; i++) {
            int left = i + 1;
            int right = n - 1;
            while(left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum >= target) {
                    right--;
                }
                else {
                    count += right - left;
                    left++;
                }
            }
        }
        return count;
    }
};
//Sean's solution
class Solution {
public:
    int threeSumSmaller(vector<int>& nums, int target) {

        sort(nums.begin(),nums.end());

        int count = 0;
        for (int i = 0; i < n - 2; i++)
            for (int j = i+1, k = n-1; j < k;)
            {
                if (nums[i] + nums[j] + nums[k] >= target)
                    k--;
                else
                {
                    count += k - j;
                    j++;
                }
            }
        return count;
    }
};